src/lib/support/DefaultStorageKeyAllocator.h

```diff
@@ -18,16 +18,68 @@
 
 #include <lib/core/CHIPPersistentStorageDelegate.h>
 #include <lib/core/DataModelTypes.h>
-#include <lib/core/GroupId.h>
-#include <lib/support/EnforceFormat.h>
-#include <lib/support/logging/Constants.h>
-#include <stdarg.h>
+#include <lib/support/CHIPMemString.h>
+
 #include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
 
 namespace chip {
 
+/**
+ * Represents a key used for addressing a specific storage element.
+ *
+ * May contain generic fixed keys (e.g. "g/fidx") or formatted fabric-specific
+ * keys ("f/%x/..." where %x is the fabric index).
+ */
+class StorageKey
+{
+public:
+    StorageKey(const StorageKey & other) = default;
+    StorageKey & operator=(const StorageKey & other) = default;
+
+    ~StorageKey() { memset(mKeyNameBuffer, 0, sizeof(mKeyNameBuffer)); }
+
+    const char * KeyName() const { return mKeyNameBuffer; }
+
+    bool IsInitialized() const { return mKeyNameBuffer[0] != 0; }
+    bool IsUninitialized() const { return mKeyNameBuffer[0] == 0; }
+    bool operator!() const { return IsUninitialized(); }
+
+    operator const char *() const { return mKeyNameBuffer; }
+
+    static StorageKey FromConst(const char * value)
```

Comment:
Previous code was not copying const strings, just setting the pointer which allowed compilers to optimize const string usage pretty well - I added that optimization to save up to 1kB of flash in some example. Maybe it's worth checking if the codesize can be reduced that way in your change, too.

Reply:
Do you happen to remember which example had this so I can verify? 
My thinking was:
   - we always have an array buffer allocated
   - having a `ptr = value` vs `strcopy(buffer, value)` seemed a wash: buffer always existed, value was always const.

Ideally we would have separate keys for const vs formatted values, however such a change is probably more involved. We can do it as a followup as I believe in most usages we know if a key is fixed or formatted.

Reply:
I think the code saving came from avoiding the function call to do the string copy. But as discussed on slack, it seems that there's no flash increase caused by your change (at least on our platform), so maybe that optimization is no longer relevant.